### PR TITLE
CORS-2969: Don't default to OpenShiftSDN after deprecation

### PR DESCRIPTION
**Pull request by zaneb**
OpenShiftSDN is being deprecated in 4.15, so switch the default to OVNKubernetes in the IPv4-only case (OVNk is already required for IPv6/dual-stack) to avoid installer failures.

This follows up on 8537fdd1e01fa07caa7f09ed03c6beafdacb9378, which sets OpenShiftSDN in the manifests (but not the install-config) iff it is *explicitly* requested.

### CODE REVIEW

**Comment by danwinship** [comment]
"Fixes #1613"

But https://github.com/openshift/release/pull/46286 and https://github.com/openshift/release/pull/46338 should probably merge before this, to ensure that no CI jobs unexpectedly switch network types when the default changes.

File: ocp_install_env.sh [hunk 1]
```diff
@@ -258,10 +258,15 @@ EOF
   fi
 }
 
+function openshift_sdn_deprecated() {
+  # OpenShiftSDN is deprecated in 4.15 and later
+  [[ "${NETWORK_TYPE}" = "OpenShiftSDN" ]] && printf '4.15\n%s\n' "$(openshift_version)" | sort -V -C
+}
```
Comment (danwinship) [inline ocp_install_env.sh:264]: `openshift_sdn_deprecated` makes it sound like this just checks if openshift-sdn is deprecated, but it actually checks if sdn is deprecated _and_ it's the selected `NETWORK_TYPE`.

**Review by danwinship — COMMENTED** [review]
This makes the control flow really weird... it still defaults to `NETWORK_TYPE=OpenShiftSDN` for IPv4 in `network.sh`, but then just overrides it later...

**Comment by zaneb** [comment]
@danwinship Fixed the control flow, hopefully this actually works for the agent as well.

> But https://github.com/openshift/release/pull/46286 and https://github.com/openshift/release/pull/46338 should probably merge before this, to ensure that no CI jobs unexpectedly switch network types when the default changes.

Can't we merge this and abandon all of https://github.com/openshift/release/pull/46563, https://github.com/openshift/release/pull/46286, and https://github.com/openshift/release/pull/46338?

From my reading of the commit messages, there's a bunch of jobs that don't have -sdn- in the title where we don't actually care what network type they have. With this change they'll carry on as they always have on existing release branches, but switch to ovnk on new ones. It seems to me that that's what we want, and it avoids renaming everything in the world (multiple times).

File: network.sh [hunk 2]
```diff
@@ -111,6 +111,11 @@ else
   exit 1
 fi
 
+function openshift_sdn_deprecated() {
+  # OpenShiftSDN is deprecated in 4.15 and later
+  printf '4.15\n%s\n' "$(openshift_version)" | sort -V -C
```
Comment (bfournie) [inline network.sh:116]: Doesn't `function openshift_version()` need to be included in network.sh for this to work?

**Comment by danwinship** [comment]
In _some_ cases it definitely makes sense to convert the SDN job to OVN-K (eg, the "virtualmedia" jobs, https://github.com/openshift/release/pull/46338).

But there are some places where there are two jobs that are completely identical except that one of them is sdn and one is ovn, so the right fix is to drop the sdn job. (I'm not sure if that case applies to any dev-scripts-based jobs though.)

The trickier case is where there are two jobs that are _nearly_ identical, like `e2e-metal-ipi-sdn` (-ipv4) vs `e2e-metal-ipi-ovn-ipv6`, where it's not clear if we still care about testing both IPv4 and IPv6 if we're not also testing both network plugins. And that seemed like a decision that was better to leave up to the individual repo owners to figure out.

> and it avoids renaming everything in the world (multiple times).

https://github.com/openshift/release/pull/46563 and https://github.com/openshift/release/pull/46286 don't rename any _jobs_; they just change job _configs_ to make it clearer when a job ends up using sdn.

(https://github.com/openshift/release/pull/46338 does rename jobs, but that's because it's actually flipping jobs from sdn to ovn.)

**Comment by danwinship** [comment]
> And that seemed like a decision that was better to leave up to the individual repo owners to figure out.

(of course, you _are_ "the individual repo owners" for some of these jobs, so if you want to figure that now, go ahead...)

File: network.sh (hunk 2 above)
Comment (bfournie) [reply network.sh:116]: Maybe its related to the ordering of the sources in` agent/06_agent_create_cluster.sh`, I had to define `openshift_version()` in network.sh in order to successfully run `make agent`

**Comment by andfasano** [comment]
/hold

Currently there are two separate issues (but somehow related) requiring to be addressed:

1. As [previously mentioned](https://github.com/openshift-metal3/dev-scripts/pull/1622#discussion_r1434140781) by @bfournie, the `openshift_version` function is not defined in `network.sh` and so it's not found
2. The current dev-scripts CI configuration was still pointing at [4.14](https://github.com/openshift/release/blob/273f642cef6ec841454c89a9f3ee37fd88fe1c65/ci-operator/config/openshift-metal3/dev-scripts/openshift-metal3-dev-scripts-master.yaml#L22), thus making pass the current builds. I'll open a new PR to fix this point

**Comment by andfasano** [comment]
Requires https://github.com/openshift/release/pull/47487

File: network.sh (hunk 2 above)
Comment (bfournie) [reply network.sh:116]: Recommend moving `openshift_release_version()` to common.sh, that should take care of the ordering issue.

**Comment by danwinship** [comment]
> But [openshift/release#46286](https://github.com/openshift/release/pull/46286) and [openshift/release#46338](https://github.com/openshift/release/pull/46338) should probably merge before this, to ensure that no CI jobs unexpectedly switch network types when the default changes.

done

> "Fixes #1613"

(as in, "you should add that to the initial comment so it closes that issue when this merges")

**Comment by andfasano** [comment]
/hold cancel

Now the agent job is failing as expected

**Comment by zaneb** [comment]
The dependencies here are much nastier than I realised, and I'm almost ready to give up on this change.
This whole pile of bash is completely unmaintainable :cry:

**Pull request merged**